user_tools/src/spark_rapids_tools/api_v1/builder.py

```diff
@@ -924,3 +930,200 @@ def combine_reports(
+        raise_on_failure: Optional[bool] = field(default=True)
+        raise_on_empty: Optional[bool] = field(default=False)
+        res_container: Optional[LoadCombinedRepResult] = field(default=None)
+        _csv_reports: List[CSVReport] = field(default_factory=list, init=False)
+        _combiner: Optional[CSVReportCombiner] = field(default=None, init=False)
+
+        @cached_property
+        def res_handlers(self) -> List[ToolResultHandlerT]:
+            """
+            Get the list of result handlers.
+            :return: A list of ToolResultHandlerT instances.
+            """
+            if isinstance(self.handlers, list):
+                return self.handlers
+            return [self.handlers]
+
+        def _init_reports(self) -> None:
+            """
+            Initialize the CSVReport instances for each result handler.
+            :return: None. The CSVReport instances are stored in the _csv_reports attribute.
+            """
+            if not self.res_handlers:
+                raise ValueError('No result handlers provided for CSVReportCombiner.')
+            # create a CSVReport for each result handler
+            for res_h in self.res_handlers:
+                self._csv_reports.append(
+                    CSVReport(res_h, _tbl=self.table)
+                )
+
+        def __enter__(self) -> 'APIHelpers.CombinedDFBuilder':
+            """
+            Context manager entry point to start building the CSVReportCombiner.
+            It initializes the CSVReports for each result handler.
+            :return: The builder instance for chaining.
+            """
+            self._init_reports()
+            return self
+
+        def __exit__(self, exc_type, exc_val, exc_tb):
+            """
+            Propagate the exception if any occurred during the context.
+            """
+            return False
+
+        ###################################
+        # Public API for API Helpers
+        ###################################
+
+        def apply_on_report(self, csv_rep_cb: Callable[[CSVReport], CSVReport]) -> 'APIHelpers.CombinedDFBuilder':
+            """
+            Apply a callBack on the CSVReport to configure it.
+            This is useful to apply any custom configuration on each CSVReport before building the combiner.
+            Note that the callback is applied on each CSVReport instance created for each result handler.
+            :param csv_rep_cb: A callback that takes a CSVReport and returns a configured CSVReport.
+                               Example of inlined callback:
+                               lambda rep: rep.pd_args({'usecols': ['col1', 'col2']})
+            :return: The builder instance for chaining.
+            """
+            if not self.res_handlers:
+                raise ValueError('No result handlers provided for CSVReportCombiner.')
+            for rep in self._csv_reports:
+                csv_rep_cb(rep)
+            return self
```

Comment:
Curious if this code can be bundled into `_init_reports` above somehow, so it's all configured in the context manager setup.  Unfortunately, it might make the context manager construction more complicated, so not super important.